# Data Cleaning

## import libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os
os.chdir("/content/drive/MyDrive/Colab Notebooks")


## Import the dataset

In [ ]:
df = pd.read_csv("data1.csv")

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (6) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [ ]:
# converting to date type
df['Date'] = pd.to_datetime(df['Date'])

In [ ]:
df = df.drop('Volume Sold (Gallons)', axis=1)

In [ ]:
# check for missing values
df.isnull().sum()

Invoice/Item Number         0
Date                        0
Store Number                0
Store Name                  0
Address                  2376
City                     2375
Zip Code                 2420
Store Location           2375
County Number           79178
County                  79178
Category                 8020
Category Name           16086
Vendor Number               3
Vendor Name                 1
Item Number                 0
Item Description            0
Pack                        0
Bottle Volume (ml)          0
State Bottle Cost          10
State Bottle Retail        10
Bottles Sold                0
Sale (Dollars)             10
Volume Sold (Liters)        0
dtype: int64

In [ ]:
# Getting rid of $ in each value of column and change the type to floats
df['State Bottle Cost'] = df['State Bottle Cost'].str.replace('$', '', regex=True).astype('float')
df['State Bottle Retail'] = df['State Bottle Retail'].str.replace('$', '', regex=True).astype('float')
df['Sale (Dollars)'] = df['Sale (Dollars)'].str.replace('$', '', regex=True).astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12591077 entries, 0 to 12591076
Data columns (total 23 columns):
 #   Column                Dtype         
---  ------                -----         
 0   Invoice/Item Number   object        
 1   Date                  datetime64[ns]
 2   Store Number          int64         
 3   Store Name            object        
 4   Address               object        
 5   City                  object        
 6   Zip Code              object        
 7   Store Location        object        
 8   County Number         float64       
 9   County                object        
 10  Category              float64       
 11  Category Name         object        
 12  Vendor Number         float64       
 13  Vendor Name           object        
 14  Item Number           int64         
 15  Item Description      object        
 16  Pack                  int64         
 17  Bottle Volume (ml)    int64         
 18  State Bottle Cost     float64       
 19

Now these columns are no longer object type, they are float.

There are also errors in the Category Name.

# Don't know if Ishould do this

In [ ]:
wrong_cat_name = ['WHISK', 'VODKA', 'TEQ', 'GIN', 'SCHNA', 'RUM', 'Liqueur', 'WINE', 'BEER', 'Bourbon', 'Brandie', 'scotch']
right_cat_name = ['WHISKEY', 'VODKA', 'TEQUILA', 'GIN', 'SCHNAPPS', 'RUM', 'LIQUEUR', 'WINE', 'BEER', 'BOURBON', 'Brandies', 'SCOTCH']

In [ ]:
def clean_cat_name(x, y):
  df.loc[df['Category Name'].str.contains(x, case=False, na=False), 'Category Name'] = y

In [ ]:
for i in range(len(wrong_cat_name)):
  clean_cat_name(wrong_cat_name[i], right_cat_name[i])

In [ ]:
print(df['Category Name'].nunique())

130


# This

In [ ]:
#df.loc[df['Store Name'] == 'Hy-Vee Food Store #2 / Iowa City']

In [ ]:
df['Store Name'].str.contains('/', case=False, na=False).sum()

8360573

In [ ]:
print(df.index[df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]].tolist())

IndexError: ignored

In [ ]:
df_store = df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]

In [ ]:
df_store

Invoice/Item Number       Date  Store Number  \
3               S28867700001 2015-11-04          2513   
6               S28865700001 2015-11-09          2538   
8               S29339300091 2015-11-30          2662   
10              S29049900001 2015-11-17          2661   
11              S28868200001 2015-11-05          2561   
...                      ...        ...           ...   
12590996     INV-08367900049 2017-10-31          2604   
12590997     INV-08367900050 2017-10-31          2604   
12590998     INV-08367900051 2017-10-31          2604   
12590999     INV-08367900052 2017-10-31          2604   
12591000     INV-08367900053 2017-10-31          2604   

                                 Store Name                Address  \
3          Hy-Vee Food Store #2 / Iowa City         812  S 1ST AVE   
6           Hy-Vee Food Store #3 / Waterloo       1422 FLAMMANG DR   
8         Hy-Vee Wine & Spirits / Muscatine  522 MULBERRY, SUITE A   
10              Hy-Vee Food Store / Sheldon           1989 PARK ST   
11          Hy-Vee Food Store / Fleur / DSM       4605 FLEUR DRIVE   
...                                     ...                    ...   
12590996   Hy-Vee Wine and Spirits / Lemars       1201 12th Ave SW   
12590997   Hy-Vee Wine and Spirits / Lemars       1201 12th Ave SW   
12590998   Hy-Vee Wine and Spirits / Lemars       1201 12th Ave SW   
12590999   Hy-Vee Wine and Spirits / Lemars       1201 12th Ave SW   
12591000   Hy-Vee Wine and Spirits / Lemars       1201 12th Ave SW   

                City Zip Code  \
3          IOWA CITY    52240   
6           WATERLOO    50702   
8          MUSCATINE    52761   
10           SHELDON    51201   
11        DES MOINES    50321   
...              ...      ...   
12590996      Lemars    51031   
12590997      Lemars    51031   
12590998      Lemars    51031   
12590999      Lemars    51031   
12591000      Lemars    51031   

                                             Store Location  County Number  \
3                          812 S 1ST AVE\nIOWA CITY 52240\n           52.0   
6         1422 FLAMMANG DR\nWATERLOO 50702\n(42.459938, ...            7.0   
8                  522 MULBERRY, SUITE A\nMUSCATINE 52761\n           70.0   
10        1989 PARK ST\nSHELDON 51201\n(43.186038, -95.8...           71.0   
11        4605 FLEUR DRIVE\nDES MOINES 50321\n(41.542816...           77.0   
...                                                     ...            ...   
12590996  1201 12th Ave SW\nLemars 51031\n(42.778257, -9...           75.0   
12590997  1201 12th Ave SW\nLemars 51031\n(42.778257, -9...           75.0   
12590998  1201 12th Ave SW\nLemars 51031\n(42.778257, -9...           75.0   
12590999  1201 12th Ave SW\nLemars 51031\n(42.778257, -9...           75.0   
12591000  1201 12th Ave SW\nLemars 51031\n(42.778257, -9...           75.0   

              County  ...                        Vendor Name Item Number  \
3            Johnson  ...                    Jim Beam Brands         237   
6         Black Hawk  ...  Duggan's Distillers Products Corp         238   
8          Muscatine  ...                    Jim Beam Brands         173   
10           O'Brien  ...        Disaronno International LLC         249   
11              Polk  ...                    Jim Beam Brands         237   
...              ...  ...                                ...         ...   
12590996    Plymouth  ...               FIFTH GENERATION INC       38174   
12590997    Plymouth  ...                  Phillips Beverage       41703   
12590998    Plymouth  ...                  PERNOD RICARD USA       42712   
12590999    Plymouth  ...               SAZERAC COMPANY  INC       64858   
12591000    Plymouth  ...              SAZERAC NORTH AMERICA       84172   

                                       Item Description Pack  \
3                        Knob Creek w/ Crystal Decanter    3   
6                          Forbidden Secret Coffee Pack    6   
8                           Laphroaig w/ Whi

# Amazing

In [ ]:
df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]['City'].isnull().sum()

1842

There are 1842 missing data in column 'City' that I can take from the column 'Store Name' where there is a slash.

In [ ]:
df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]['City'] is np.nan

False

In [ ]:
for i in range(8360574):
  if df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]['City'].isnull() == True:
    print(df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True])

KeyError: ignored

In [ ]:
ind = df.index[df_store].tolist()

IndexError: ignored

In [ ]:
df.shape[0]

12591077

In [ ]:
df_store.where(cond=df_store['City'].isna(), inplace = False)

In [ ]:
for df_store in df:
  if df_store['City'].isnull() == True:
    df.loc[df['Store Name'] == df_store['Store Name'][df_store], ['Store Name', 'City']] = df_store.split('/')
  else:
    pass

TypeError: ignored

In [ ]:
for i in df_store:
  if df_store['City'].isnull() == True:
    df.loc[df['Store Name'] == df_store['Store Name'][i], ['Store Name', 'City']] = i.split('/')

ValueError: ignored

In [ ]:
df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]['Store Name'][3]

'Hy-Vee Food Store #2 / Iowa City'

In [ ]:
df.iloc[11]

Invoice/Item Number                                          S28868200001
Date                                                  2015-11-05 00:00:00
Store Number                                                         2561
Store Name                                Hy-Vee Food Store / Fleur / DSM
Address                                                  4605 FLEUR DRIVE
City                                                           DES MOINES
Zip Code                                                            50321
Store Location          4605 FLEUR DRIVE\nDES MOINES 50321\n(41.542816...
County Number                                                        77.0
County                                                               Polk
Category                                                              NaN
Category Name                                                         NaN
Vendor Number                                                        65.0
Vendor Name                           

In [ ]:
df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True]

In [ ]:
if df['Store Name'].str.contains('/', case=False, na=False) == True:
  df.loc[df['Store Name'] == x, ['Store Name', 'City']] = x.split('/')

In [ ]:
for x in df.loc[df['Store Name'].str.contains('/', case=False, na=False) == True][3]:
  df.loc[df['Store Name'] == x, ['Store Name', 'City']] = x.split('/')

KeyError: ignored

In [ ]:
df.loc[df['Store Name'] == df['Store Name'].str.contains('/', case=False, na=False)]

Empty DataFrame
Columns: [Invoice/Item Number, Date, Store Number, Store Name, Address, City, Zip Code, Store Location, County Number, County, Category, Category Name, Vendor Number, Vendor Name, Item Number, Item Description, Pack, Bottle Volume (ml), State Bottle Cost, State Bottle Retail, Bottles Sold, Sale (Dollars), Volume Sold (Liters)]
Index: []

[0 rows x 23 columns]

In [ ]:
wrong_str_name = ['Hy-Vee Food Store #2 / Iowa City', 'Kum & Go 202 / 4Th St Waukee']

In [ ]:
x = 'Hy-Vee Food Store #2 / Iowa City'
df.loc[df['Store Name'] == x, ['Store Name', 'City']] = x.split('/')

In [ ]:
df.loc[df['Store Name'] == 'Hy-Vee Food Store #2 ']

Invoice/Item Number       Date  Store Number             Store Name  \
3               S28867700001 2015-11-04          2513  Hy-Vee Food Store #2    
509             S27940500043 2015-09-16          2513  Hy-Vee Food Store #2    
1097            S11874400013 2013-04-25          2513  Hy-Vee Food Store #2    
1209            S25218700012 2015-04-22          2513  Hy-Vee Food Store #2    
1291            S05247600134 2012-04-26          2513  Hy-Vee Food Store #2    
...                      ...        ...           ...                    ...   
12583693     INV-08345300155 2017-10-31          2513  Hy-Vee Food Store #2    
12583694     INV-08345300156 2017-10-31          2513  Hy-Vee Food Store #2    
12583695     INV-08345300157 2017-10-31          2513  Hy-Vee Food Store #2    
12583696     INV-08345300158 2017-10-31          2513  Hy-Vee Food Store #2    
12583697     INV-08345300159 2017-10-31          2513  Hy-Vee Food Store #2    

                 Address        City Zip Code  \
3         812  S 1ST AVE   Iowa City    52240   
509       812  S 1ST AVE   Iowa City    52240   
1097      812  S 1ST AVE   Iowa City    52240   
1209      812  S 1ST AVE   Iowa City    52240   
1291      812  S 1ST AVE   Iowa City    52240   
...                  ...         ...      ...   
12583693  812  S 1st Ave   Iowa City    52240   
12583694  812  S 1st Ave   Iowa City    52240   
12583695  812  S 1st Ave   Iowa City    52240   
12583696  812  S 1st Ave   Iowa City    52240   
12583697  812  S 1st Ave   Iowa City    52240   

                            Store Location  County Number   County  ...  \
3         812 S 1ST AVE\nIOWA CITY 52240\n           52.0  Johnson  ...   
509       812 S 1ST AVE\nIOWA CITY 52240\n           52.0  Johnson  ...   
1097      812 S 1ST AVE\nIOWA CITY 52240\n           52.0  Johnson  ...   
1209      812 S 1ST AVE\nIOWA CITY 52240\n           52.0  Johnson  ...   
1291      812 S 1ST AVE\nIOWA CITY 52240\n           52.0  Johnson  ...   
...                                    ...            ...      ...  ...   
12583693  812 S 1st Ave\nIowa City 52240\n           52.0  JOHNSON  ...   
12583694  812 S 1st Ave\nIowa City 52240\n           52.0  JOHNSON  ...   
12583695  812 S 1st Ave\nIowa City 52240\n           52.0  JOHNSON  ...   
12583696  812 S 1st Ave\nIowa City 52240\n           52.0  JOHNSON  ...   
12583697  812 S 1st Ave\nIowa City 52240\n           52.0  JOHNSON  ...   

                               Vendor Name Item Number  \
3                          Jim Beam Brands         237   
509                         Luxco-St Louis       31658   
1097                  Bacardi U.S.A., Inc.       43126   
1209      Constellation Wine Company, Inc.       11777   
1291                    Heaven Hill Brands       41316   
...                                    ...         ...   
12583693                Brown Forman Corp.       87586   
12583694                    Gemini Spirits       88036   
12583695                           PROXIMO       89194   
12583696                Brown Forman Corp.       89278   
12583697                         LUXCO INC       89387   

                             Item Description Pack  Bottle Volume (ml)  \
3              Knob Creek w/ Crystal Decanter    3                1750   
509                             Paramount Gin    6                1750   
1097                     Bacardi Superior Rum   12                 750   
1209                             Black Velvet   12                1000   
1291                 Burnetts Raspberry Vodka   12                 750   
...                                       ...  ...                 ...   
12583693            El Jimador Silver Tequila   12                 750   
12583694        Margaritaville Silver Tequila   12                 750   
12583695  Jose Cuervo Especial Reposado Flask   24                 375   
12583696          El Jimador Reposado Tequila   12                 750   
12583697                  Juarez Tequila Gold  

There are capitalization errors in the dataset.

In [ ]:
#getting rid of an error of repeating names due to difference in capitilization
df["City"]= df["City"].replace('Des Moines', "DES MOINES")
df["City"]= df["City"].replace('Cedar Rapids', "CEDAR RAPIDS")
df["City"]= df["City"].replace('Davenport', "DAVENPORT")
df["City"]= df["City"].replace('Council Bluffs', "COUNCIL BLUFFS")
df["City"]= df["City"].replace('Waterloo', "WATERLOO")
df["City"]= df["City"].replace('West Des Moines', "WEST DES MOINES")
df["City"]= df["City"].replace('Sioux City', "SIOUX CITY")
df["City"]= df["City"].replace('Dubuque', "DUBUQUE")
df["City"]= df["City"].replace('Iowa City', "IOWA CITY")
df["City"]= df["City"].replace('Ames', "AMES")

In [ ]:
df.isnull().sum()

Invoice/Item Number         0
Date                        0
Store Number                0
Store Name                  0
Address                  2376
City                     2375
Zip Code                 2420
Store Location           2375
County Number           79178
County                  79178
Category                 8020
Category Name           16086
Vendor Number               3
Vendor Name                 1
Item Number                 0
Item Description            0
Pack                        0
Bottle Volume (ml)          0
State Bottle Cost          10
State Bottle Retail        10
Bottles Sold                0
Sale (Dollars)             10
Volume Sold (Liters)        0
dtype: int64

In [ ]:
df['Vendor Number'].astype(int)

IntCastingNaNError: ignored

# Feature Engineering